src/PostgREST/Statements.hs

```diff
@@ -77,30 +79,30 @@ createWriteStatement selectQuery mutateQuery wantSingle isInsert asCsv rep pKeys
 
   selectF
     -- prevent using any of the column names in ?select= when no response is returned from the CTE
-    | rep `elem` [None, HeadersOnly] = "SELECT * FROM " <> sourceCTEName
+    | rep `elem` [None, HeadersOnly] = H.sql ("SELECT * FROM " <> sourceCTEName)
     | otherwise                      = selectQuery
 
   decodeStandard :: HD.Result ResultsWithCount
   decodeStandard =
    fromMaybe (Nothing, 0, [], mempty, Right [], Right Nothing) <$> HD.rowMaybe standardRow
 
-createReadStatement :: SqlQuery -> SqlQuery -> Bool -> Bool -> Bool -> Maybe FieldName -> PgVersion ->
+createReadStatement :: H.Snippet -> H.Snippet -> Bool -> Bool -> Bool -> Maybe FieldName -> PgVersion ->
                        H.Statement () ResultsWithCount
 createReadStatement selectQuery countQuery isSingle countTotal asCsv binaryField pgVer =
-  H.Statement sql HE.noParams decodeStandard False
+  H.dynamicallyParameterized snippet decodeStandard True
  where
-  sql = [qc|
```

Comment:
I find those quasi-quoted statements a lot easier to read. Do we need to change those?

Reading http://hackage.haskell.org/package/interpolatedstring-perl6-1.0.2/docs/Text-InterpolatedString-Perl6.html I expect this to work as well:

```
snippet = [qc|
  WITH
  {H.sql sourceCTEName} AS ({selectQuery})
...
```

Coolest thing would be to have a `[sql|` quasi-quoter that takes snippets and strings and returns a snippet or something like that?

Reply:
There seems to be `[uncheckedSql|` here: https://hackage.haskell.org/package/hasql-th-0.4.0.8/docs/Hasql-TH.html#g:4

Not sure if that helps with DynamicStatements though.

Reply:
Yes, I agree. They were easier to read. Also they looked better on pg logs.

I changed them because interpolating the `Snippet` didn't work. We also had a discussion a long time ago about removing `qc` because it had a perf impact(didn't confirm).

> Coolest thing would be to have a [sql| quasi-quoter that takes snippets and strings and returns a snippet or something like that?

Looks possible! I'll check that out.

Reply:
@wolfgangwalther What do you think about leaving the refactor for another PR? I'd like to merge and see if I can improve #1600 with hasql-dynamic.

We'll be definitely working on the SQL fragments and we'll notice if it's hard to work with these new Snippets.

Reply:
> What do you think about leaving the refactor for another PR? I'd like to merge and see if I can improve #1600(the SET LOCALs) with hasql-dynamic.

Yes, of course.

> We'll be definitely working on the SQL fragments and we'll notice if it's hard to work with these new Snippets.

If this also makes them less nice to look at, we might want to do something about it for #1578. I'm sure we'll pick up on this.